subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1534,6 +1525,34 @@ private void assertIsResolvable() {
         }
     }
 
+    private void assertIsDeclarable() {
+        if (!canBeDeclared) {
+            throw new IllegalStateException("Declaring dependencies for configuration '" + name + "' is not allowed as it is defined as 'canBeDeclared=false'.\n");
+        }
+    }
+
+    /**
+     * Ensure that the combination of consumable, resolvable and declarable flags is sensible.
+     * This method should be called only after all mutations are known to be complete.
+     */
+    private void assertIsProperConfiguration() {
+        if (canBeConsumed && canBeResolved) {
+            DeprecationLogger.deprecateBehaviour("The configuration " + identityPath.toString() + " is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.")
+                    .willBecomeAnErrorInGradle9()
+                    .withUserManual("BLAH", "TODO: LOOK ME UP")
+                    .nagUser();
+        }
+
+        if (canBeConsumed && canBeDeclared) {
+            DeprecationLogger.deprecateBehaviour("The configuration " + identityPath.toString() + " is both consumable and declarable. This combination is incorrect, only one of these flags should be set.")
+                    .willBecomeAnErrorInGradle9()
+                    .withUserManual("BLAH", "TODO: LOOK ME UP")
+                    .nagUser();
+        }
+
+        // canBeDeclared && canBeResolved is a valid and expected combination
```

Comment:
Curious if we plan to continue supporting this in the future when we do the big `split configurations into 3 different types` thing.